demos/common/utilities/azure_sample_ca_recovery.c

```diff
@@ -0,0 +1,241 @@
+                     AZ_SPAN_FROM_STR( AZ_IOT_CA_RECOVERY_EXPIRY_TIME_NAME ) ) )
+        {
+            _az_RETURN_IF_FAILED( az_json_reader_next_token( ref_json_reader ) );
+            RETURN_IF_JSON_TOKEN_NOT_TYPE( ref_json_reader, AZ_JSON_TOKEN_STRING );
+
+            if( ref_json_reader->token.kind != AZ_JSON_TOKEN_NULL )
+            {
+                trust_bundle->expiry_time = ref_json_reader->token.slice;
+            }
+        }
+        else if( az_json_token_is_text_equal(
+                     &ref_json_reader->token,
+                     AZ_SPAN_FROM_STR( AZ_IOT_CA_RECOVERY_CERTS_NAME ) ) )
+        {
+            _az_RETURN_IF_FAILED( az_json_reader_next_token( ref_json_reader ) );
+            RETURN_IF_JSON_TOKEN_NOT_TYPE( ref_json_reader, AZ_JSON_TOKEN_STRING );
+
+            if( ref_json_reader->token.kind != AZ_JSON_TOKEN_NULL )
+            {
+                trust_bundle->certificates = ref_json_reader->token.slice;
+            }
+        }
+
+        _az_RETURN_IF_FAILED( az_json_reader_next_token( ref_json_reader ) );
+    }
+
+    return AZ_OK;
+}
+
+static az_result az_iot_ca_recovery_parse_recovery_payload( az_json_reader * ref_json_reader,
+                                                            azure_iot_ca_recovery_payload * recovery_payload )
+{
+    /* _az_PRECONDITION_NOT_NULL(ref_json_reader); */
+    /* _az_PRECONDITION_NOT_NULL(recovery_payload); */
+
+    _az_RETURN_IF_FAILED( az_json_reader_next_token( ref_json_reader ) );
+    RETURN_IF_JSON_TOKEN_NOT_TYPE( ref_json_reader, AZ_JSON_TOKEN_BEGIN_OBJECT );
+    _az_RETURN_IF_FAILED( az_json_reader_next_token( ref_json_reader ) );
+
+    recovery_payload->iothub_hostname = AZ_SPAN_EMPTY;
+    recovery_payload->payload_signature = AZ_SPAN_EMPTY;
+    recovery_payload->trust_bundle_json_object_text = AZ_SPAN_EMPTY;
+
+    while( ref_json_reader->token.kind != AZ_JSON_TOKEN_END_OBJECT )
+    {
+        RETURN_IF_JSON_TOKEN_NOT_TYPE( ref_json_reader, AZ_JSON_TOKEN_PROPERTY_NAME );
+
+        if( az_json_token_is_text_equal(
+                &ref_json_reader->token,
+                AZ_SPAN_FROM_STR( AZ_IOT_CA_RECOVERY_HUB_HOSTNAME_NAME ) ) )
+        {
+            _az_RETURN_IF_FAILED( az_json_reader_next_token( ref_json_reader ) );
+            RETURN_IF_JSON_TOKEN_NOT_TYPE( ref_json_reader, AZ_JSON_TOKEN_STRING );
+
+            if( ref_json_reader->token.kind != AZ_JSON_TOKEN_NULL )
+            {
+                recovery_payload->iothub_hostname = ref_json_reader->token.slice;
+            }
+        }
+        else if( az_json_token_is_text_equal(
+                     &ref_json_reader->token,
+                     AZ_SPAN_FROM_STR( AZ_IOT_CA_RECOVERY_PAYLOAD_NAME ) ) )
+        {
+            _az_RETURN_IF_FAILED( az_json_reader_next_token( ref_json_reader ) );
+            RETURN_IF_JSON_TOKEN_NOT_TYPE( ref_json_reader, AZ_JSON_TOKEN_BEGIN_OBJECT );
+
+            while( ref_json_reader->token.kind != AZ_JSON_TOKEN_END_OBJECT )
```

Comment:
Nit: create a separate function (e.g. inline in this case) to lower code complexity.